pkg/controller/unreachable/unreachable_controller.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright (C) 2019 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package unreachable provides a controller which periodically checks if a remote cluster is reachable
+// and maintains a condition on the cluster as a result. If the unreachable condition is true, other controllers
+// can skip attempts to reach the cluster which require a 30 second timeout.
+package unreachable
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+const (
+	controllerName = "unreachable"
+
+	adminKubeConfigKey                 = "kubeconfig"
+	maxUnreachableDuration             = 2 * time.Hour
+	maxNoOfAttemptsForUnreachableCheck = 4
```

Comment:
Confused about this vars naming, I thought it was just a made up value to get a reasonable re-try period, not a max number of attempts. Not an easy one to name if so.

Reply:
Agree, will fix the variable name.